Allow specifying custom request headers.

thatch opened the pull request:
When used with a CacheControl session, this allows requesting fresh data with `{"Cache-Control": "max-age=0"}` which is what pip does:

https://github.com/pypa/pip/blob/51de88ca6459fdd5213f86a54b021a80884572f9/src/pip/_internal/index/collector.py#L145-L158

I spent a couple of minutes trying to make a test in `test/test_client.py` for this but got confused pretty quickly because they all failed with 500's after I added what I thought would work.  Any advice (or could you show me in a subsequent commit)?

jwodder commented on src/pypi_simple/client.py:
```diff
@@ -4,7 +4,7 @@
 import platform
 from types import TracebackType
-from typing import Any, AnyStr, Optional
+from typing import Any, AnyStr, Dict, Optional
```
This file has `from __future__ import annotations`, so you can just use `dict` instead of `typing.Dict` (and I should probably replace the `Optional[T]`'s with `T | None` at some point too...).

jwodder commented on src/pypi_simple/client.py:
```diff
@@ -113,12 +114,18 @@ def get_index_page(
 
             ``accept`` parameter added
 
+        .. versionchanged:: 1.4.2
```
This is a new feature, so the next version will actually be 1.5.0.

jwodder commented on src/pypi_simple/client.py:
```diff
@@ -113,12 +114,18 @@ def get_index_page(
             The :mailheader:`Accept` header to send in order to
             specify what serialization format the server should return;
             defaults to the value supplied on client instantiation
+        :param Optional[Dict[str, str]] headers:
```
```suggestion
        :param Optional[Dict[str,str]] headers:
```

I believe Sphinx requires the type in a `:param:` to be a single word.

thatch replied:
I ran `tox -e docs` and it seems to be fine with this (after `s/Dict/dict/`)?

![image](https://github.com/jwodder/pypi-simple/assets/49834/ddadd5e5-d3e1-46f9-877a-274f12629540)

jwodder replied:
Huh.  Never mind, then.

jwodder reviewed (changes requested):
Please add this to all the other request methods as well (i.e., `stream_project_names()`, `download_package()`, and `get_package_metadata()`).

jwodder commented:
Regarding tests, something like the following should work:

```python
@responses.activate
def test_custom_headers() -> None:
    with (DATA_DIR / "simple01.html").open() as fp:
        responses.add(
            method=responses.GET,
            url="https://test.nil/simple/",
            body=fp.read(),
            content_type="text/html",
            match=responses.matchers.header_matcher({"X-Custom": "foo"}),
        )
    with PyPISimple("https://test.nil/simple/") as simple:
        # Just check that the method returns successfully
        simple.get_index_page(headers={"X-Custom": "foo"})
```

thatch commented:
> Please add this to all the other request methods as well (i.e., `stream_project_names()`, `download_package()`, and `get_package_metadata()`).

I added it to `stream_project_names` but I'm struggling to figure out a use case for cache headers in the other two (since they're immutable resources).  Are there other headers that are worth sending?

jwodder commented:
I'm not aware of any headers that someone might currently want to sent to those endpoints, but it's conceivable that a non-PyPI implementation of the Simple API may makes use of extra headers for some purpose, and I don't think a lack of imagination on our part should stop us from giving the request methods consistent APIs.

thatch commented:
I think I modified all the methods, and added (cursory) tests for the ones that increased the coverable lines.

This is the last change I have in the queue, btw.

jwodder commented:
This is now released in pypi-simple 1.5.0.  Thank you for the contribution!
